apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1133 @@
+can create their own provider, and every package can declare which providers
+they precisely need — and installers automatically pull them in as necessary.
+
+Now, this kind of flexibility also required a slightly different approach
+than tags. After all, a single variant can be described using a number
+of properties: CUDA or ROCm runtime version, compatible GPU architectures,
+compatible CPU architectures, required instruction sets, selected libraries.
+While some variants could be described using tag-style short strings, making
+them fully scalable required using more general property lists.
+
+Expressing these properties in a key-value form seemed like a good idea.
+That is, rather than having a dozen property tags such as `cuda_lower_12_5`,
+`cuda_lower_12_6`, `cuda_lower_12_7` and so on, we'd rather have a single
+feature name `cuda_version_lower_bound` and the version as a value. And since
+we expect variant providers to be developed largely independently, putting
+these features in namespaces makes it possible to organize them logically
+and avoid name collisions. Consider the following example.
+
+<figure>
+
+```rust
+nvidia :: cuda_version_lower_bound :: 12.6
+nvidia :: cuda_version_upper_bound :: 13
+x86_64 :: level :: v3
+x86_64 :: sha_ni :: on
+```
+
+<figcaption>Listing 1. Example wheel variant properties</figcaption>
+</figure>
+
+Yes, the syntax was inspired by [trove classifiers](https://pypi.org/classifiers/)!
+Every property is a 3-element tuple consisting of the namespace (one per plugin
+used), feature name and value.
+
+Here, two providers are used: one representing CUDA installation (`nvidia`
+namespace) and one representing x86_64 CPU properties (`x86_64` namespace).
+
+<div style={{ textAlign: "center" }}>
+<figure style={{width: 'auto', margin: '0 2em', display: 'inline-block', verticalAlign: 'top'}}>
+  <img src="/posts/python-wheels-from-tags-to-variants/property-examples.png"
+    width="761" height="313" alt="A diagram showing example properties from two plugins.
+The 'nvidia' plugin provides 'cuda_version_lower_bound' with values of 12.8, 12.7, 12.6, 12.5, …
+and 'cuda_version_upper_bound' with values of …, 12.12, 12.11, 12.10, 12.9. The 'x86_64' plugin
+provides a level property with values of v3, v2 and v1, and a bunch of instruction set
+properties named 'aes', 'pclmuldqd', 'rdseed', 'sha_ni' — all of them having the value 'on'." />
+  <figcaption>Fig. 5. Example supported properties reported by plugins</figcaption>
+</figure>
+</div>
+
+The `nvidia` provider detects the local CUDA installation and determines which version
+is installed. Then, it transforms this version into set of compatible
+constraints that can be used to determine whether a given wheel variant
+can be installed — and if multiple variants fit, which one should be used.
+
+For example, if you have CUDA 12.8, then all wheels with lower bound that
+is not higher than that are installable — but a wheel built specifically
```

Comment:
"all wheels with lower bound that is not higher than that are installable" is pretty confusing. I think you should try to rephrase that in a simpler way.

Reply:
Yeah, I think "that do not require a newer CUDA version" will be clearer.